### DISCUSSION TITLE
Not converge in liquid-metal calculation when velocity is low

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#25315] [Q&A General]
@joshuahansel @GiudGiud Hi guys, I create the subchannel calculation app on the basis of THM. It calculates all the channels together. But I found the  convergence is not good when the velocity is low(about 0.1 m/s).  As far as I know, SIMPLE method also has this problem. But SIMPLE can solve the case whose velocity is about 0.01 m/s(can't be lower). In MOOSE, the solving method is PJFNK. I'm wondering if there is some numerical methods in MOOSE which can help and solve this problem.

Ethan

**Comment by joshuahansel** [comment]
Have you already tried NEWTON instead of PJFNK?

I'm not sure why convergence would be poor for small velocities with the spatial discretization in THM - I have not observed this before.

**Reply by Ethan-xj** [reply]
> Have you already tried NEWTON instead of PJFNK?

Thanks. I'll have a try.

> I'm not sure why convergence would be poor for small velocities with the spatial discretization in THM - I have not observed this before.

Not in THM, but in my own subchannel app. In THM, I've tried a small velocity case, there's no non-converged situation.

**Reply by GiudGiud** [reply]
If it s based on THM it might struggle too. Sodium is nearly incompressible right?

**Reply by Ethan-xj** [reply]
Yes, nearly incompressible. Oh I remember something, I change the numerical flux function from `ADNumericalFlux3EqnHLLC` to `ADNumericalFlux3EqnCentered` to calculate incompressible fluid. `ADNumericalFlux3EqnCentered` is not steady and it might be the reason. Maybe we can find a HLLC method for incompressible fluid?

**Reply by joshuahansel** [reply]
Right `ADNumericalFlux3EqnCentered` is not recommended because it is not stable; it is recommended for testing only. I don't think HLLC makes sense for a truly incompressible fluid because it's based on finite wave speeds, but you may just be dealing with _nearly_ incompressible fluids. `SodiumSaturationFluidProperties` for example could be an issue without any pressure dependence. I'm not sure. For _nearly_ incompressible, HLLC will technically work, but have a lot of discretization error, so you may need to use a lot of mesh refinement, depending on your situation.

**Reply by Ethan-xj** [reply]
Thanks. First I need to find an available HLLC method for incompressible fluids. Btw, have you ever found any info about it?
And I have another question: Is slope limiter necessary? If I don't add slope limiter, will the convergence be influenced?

**Reply by joshuahansel** [reply]
There are "low-Mach fixes" for HLLC that attempt to dial back some of the discretization error. I've tried one before and had limited success - it's problem-dependent. For slope reconstruction, I'd just do `slope_reconstruction = FULL` unless you see some kind of oscillatory results. For the problems we typically run, I rarely see any issues. I sometimes try the slope _limiter_ options (e.g., minmod), but I often see bad results, except for the toy compressible flow problems like shock tubes.

**Reply by Ethan-xj** [reply]
@joshuahansel I find that there's some mistakes in my wall friction model. And when I use a correct model, I find the low-velocity case still don't converge using PJFNK, but converges using NEWTON. I use SMP preconditioning all the time. Is there any difference in the convergence properties between PJFNK and Newton methods?

Ethan

**Reply by GiudGiud** [reply]
Hi Ethan

can you try setting `full=true` for pjfnk in the SMP preconditioning

**Reply by Ethan-xj** [reply]
@GiudGiud I set `full=true` in SMP preconditioning all the time. But I don't know why the convergence properties between PJFNK and Newton methods is different. I mean NEWTON and PJFNK with full SMP preconditioning.

**Reply by joshuahansel** [reply]
> Is there any difference in the convergence properties between PJFNK and Newton methods?

There is certainly a difference. My opinion is: if you have the exact Jacobian (which should be the case in THM), use NEWTON. If one doesn't work, try the other.
